### PR TITLE
Fix least squares solvers

### PR DESCRIPTION
**Pull request by tachella**
[least_squares](https://deepinv.github.io/deepinv/api/stubs/deepinv.optim.utils.least_squares.html#deepinv.optim.utils.least_squares) solvers have an optional regularization parameter `gamma` that can be either None, a scalar, or a tensor. The tensor version is used in cases where the regularization is different across elements in a batch. There are multiple checks on gamma, such as [this one](https://github.com/deepinv/deepinv/blob/6d99da8d90f6ab0216b36b460415faa7677044be/deepinv/optim/utils.py#L171), which assume that `gamma` is a scalar and generate an error if it is not. This is particularly problematic since it breaks `RAM` and other solvers when deployed on a batch.

Solves #786  and #770 

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [ ] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: deepinv/optim/utils.py [hunk 1]
```diff
@@ -90,11 +91,24 @@ def least_squares(
...
         parallel_dim = [parallel_dim]
+
+    zero_gamma = False
     if gamma is None:
         gamma = 0.0
+        zero_gamma = True
+    elif isinstance(gamma, (int, float)):
+        zero_gamma = not (gamma > 0)
+    elif isinstance(gamma, Tensor):
+        zero_gamma = not torch.any(gamma > 0)
```
Comment (Andrewwango) [inline deepinv/optim/utils.py]: I would use torch.all

File: deepinv/optim/utils.py [hunk 2]
```diff
@@ -90,11 +91,24 @@ def least_squares(
...
+        if not zero_gamma:
+            gamma[gamma <= 0] = 1e8  # guardrails
```
Comment (Andrewwango) [inline deepinv/optim/utils.py]: If I understand correctly, gamma is changed on the fly where it's not positive.

- Why doesn't this modification occur for a single gamma int/float too? should it be set to eps?
- Does it make sense that `gamma = [-1, 1e-10]` -> `[1e-8, 1e-10]`? surely the threshold should be set at eps

File: deepinv/optim/utils.py [hunk 3]
```diff
@@ -90,11 +91,24 @@ def least_squares(
...
     if isinstance(parallel_dim, int):
         parallel_dim = [parallel_dim]
+
+    zero_gamma = False
     if gamma is None:
         gamma = 0.0
+        zero_gamma = True
+    elif isinstance(gamma, (int, float)):
+        zero_gamma = not (gamma > 0)
+    elif isinstance(gamma, Tensor):
```
Comment (Andrewwango) [inline deepinv/optim/utils.py]: can we allow gamma to be a list too?

File: deepinv/optim/utils.py [hunk 4]
```diff
@@ -90,11 +91,24 @@ def least_squares(
     """
     if isinstance(parallel_dim, int):
         parallel_dim = [parallel_dim]
+
+    zero_gamma = False
     if gamma is None:
         gamma = 0.0
+        zero_gamma = True
+    elif isinstance(gamma, (int, float)):
+        zero_gamma = not (gamma > 0)
```
Comment (Andrewwango) [inline deepinv/optim/utils.py]: gamma <= 0 would be clearer I think

**Comment by Andrewwango** [comment]
should we also add a test that RAM can work with batch size > 1?

File: deepinv/optim/utils.py (hunk 2 above)
Comment (tachella) [reply deepinv/optim/utils.py]: the solution is different if `gamma=0` or if `gamma != 0`. If gamma is a scalar, this is easy to check. If `gamma` is not a scalar, the solver can only work if all `gamma=0` or if all `gamma != 0`. So we shouldn't threshold by default, just handle the (rare) case of a batch of gamma where some are zero and some are not

File: deepinv/optim/utils.py (hunk 3 above)
Comment (tachella) [reply deepinv/optim/utils.py]: I don't see the application

**Comment by tachella** [comment]
> should we also add a test that RAM can work with batch size > 1?

done!

File: deepinv/tests/test_optim.py [hunk 5]
```diff
@@ -999,7 +1005,7 @@ def A_adjoint(self, y, **kwargs):
 @pytest.mark.parametrize("solver", solvers)
 def test_least_squares_implicit_backward(device, solver, physics_name):
     # Check that the backward gradient matches the finite difference gradient
-    batch_size = 1
+    batch_size = 2  # TODO: test batching
```
Comment (Andrewwango) [inline deepinv/tests/test_optim.py]: add as param and do batch size 1 and 2?

File: deepinv/tests/test_optim.py [hunk 6]
```diff
@@ -1016,7 +1022,9 @@ def test_least_squares_implicit_backward(device, solver, physics_name):
     # Check gradients w.r.t y, z, gamma
     y.requires_grad_(True)
     z = torch.randn_like(x).requires_grad_(True)
-    gamma = torch.tensor(0.4, dtype=dtype, requires_grad=True)
+    gamma = (
```
Comment (Andrewwango) [inline deepinv/tests/test_optim.py:1025]: I think it would be good to test different gammas since the logic is quite nuanced + can change the function a lot!

File: deepinv/tests/test_models.py [hunk 7]
```diff
@@ -906,7 +906,10 @@ def test_restoration_models(
         if physics is not None:
             physics.noise_model = dinv.physics.GaussianNoise(0.01, rng=rng)
 
-    x = DummyCircles(imsize=imsize, samples=1)[0].unsqueeze(0)
+    x = DummyCircles(imsize=imsize, samples=1)[0].unsqueeze(0).to(device)
+
+    # make batch with > 1 element to test batch processing
+    x = x.repeat(2, 1, 1, 1)
```
Comment (Andrewwango) [inline deepinv/tests/test_models.py]: can we test with two different images in a batch, and test that both individually are restored? to make sure the images in a batch don't affect each other

File: deepinv/models/ram.py [hunk 8]
```diff
@@ -175,6 +187,8 @@ def realign_input(self, x, physics, y, sigma):
         snr = num / (sigma + 1e-4)  # SNR equivariant
         gamma = 1 / (1e-4 + 1 / (snr * f**2))
         gamma = gamma[(...,) + (None,) * (x.dim() - 1)]
+        gamma = gamma * self.fact_realign
```
Comment (Andrewwango) [inline deepinv/models/ram.py:189]: Given how brittle this was and how long it took you to find that Hai's PR had broken RAM, I think we should do more solid unit tests for RAM e.g. for different gammas, differently scaled inputs (e.g. butterfly / 1000, butterfly * 1000) to catch big things that might break it

File: deepinv/models/ram.py [hunk 9]
```diff
@@ -37,6 +37,18 @@ class RAM(Reconstructor, Denoiser):
...
     :param bool, str pretrained: If `True`, the model will be initialized with pretrained weights. If `str`, load from file.
     :param float sigma_threshold: Threshold (minimum value) for the noise level. Default is 1e-3.
+
+    |sep|
+
+      >>> import deepinv as dinv
+      >>> x = dinv.utils.load_example("butterfly.png")
+      >>> physics = dinv.physics.Downsampling(filter="bicubic")
+      >>> y = physics(x)
+      >>> model = dinv.models.RAM(pretrained=True) # or any of the models listed below
```
Comment (Andrewwango) [inline deepinv/models/ram.py]:
```suggestion
      >>> model = dinv.models.RAM(pretrained=True)
```

File: deepinv/optim/utils.py [hunk 10]
```diff
@@ -78,7 +87,7 @@ def least_squares(
     :param torch.Tensor y: input tensor of shape (B, ...)
     :param torch.Tensor z: input tensor of shape (B, ...) or scalar.
     :param torch.Tensor init: (Optional) initial guess for the solver. If None, it is set to a tensor of zeros.
-    :param None, float gamma: (Optional) inverse regularization parameter.
+    :param None, float, torch.Tensor gamma: (Optional) inverse regularization parameter. Can be batched (shape (B, ...)) or a scalar. If None, it is set to :math:`\infty` (no regularization).
     :param str solver: solver to be used.
```
Comment (Andrewwango) [inline deepinv/optim/utils.py]: list different solvers or ref where these solves are listed?

File: deepinv/optim/utils.py [hunk 11]
```diff
@@ -90,11 +99,39 @@ def least_squares(
     """
     if isinstance(parallel_dim, int):
         parallel_dim = [parallel_dim]
+
     if gamma is None:
-        gamma = 0.0
+        gamma = torch.tensor(0.0, device=y.device)
+        no_gamma = True
```
Comment (Andrewwango) [inline deepinv/optim/utils.py]: could you rename this like `gamma_provided=False` or something to avoid double negatives?

File: deepinv/optim/utils.py [hunk 12]
```diff
@@ -90,11 +99,39 @@ def least_squares(
...
+
+        if torch.any(gamma <= 0):
+            warnings.warn(
+                "Regularization parameter of least squares problem (gamma) should be positive."
+                "Otherwise, the problem can become non-convex and the solvers are not designed for that."
+                "Continuing anyway..."
+            )
+    if gamma.ndim > 0:  # if batched gamma
+        if isinstance(y, TensorList):
+            batch_size = y[0].size(0)
```
Comment (Andrewwango) [inline deepinv/optim/utils.py:120]: maybe use the `shape` property of tensorlist? i.e. `y.shape[0][0]` - do you think in tensorlist we should enforce that all tensors inside the tensorlist have the same batch size?

File: deepinv/optim/utils.py [hunk 13]
```diff
@@ -370,6 +407,7 @@ def _sym_ortho(a, b):
     ``1/eps`` in some important places.
 
     """
+    a, b = torch.broadcast_tensors(a, b)
```
Comment (Andrewwango) [inline deepinv/optim/utils.py:410]: nice, I have never ever heard of broadcast_tensors before haha

File: deepinv/optim/utils.py (hunk 12 above)
Comment (tachella) [reply deepinv/optim/utils.py:120]: I think size is better as it returns a Tensor - and yes we could enforce this potentially, but I would leave it for other PR

File: deepinv/optim/utils.py (hunk 11 above)
Comment (tachella) [reply deepinv/optim/utils.py]: done

File: deepinv/optim/utils.py (hunk 10 above)
Comment (tachella) [reply deepinv/optim/utils.py]: done

File: deepinv/tests/test_optim.py (hunk 6 above)
Comment (tachella) [reply deepinv/tests/test_optim.py:1025]: I'm randomizing to make sure that it works

File: deepinv/tests/test_models.py (hunk 7 above)
Comment (tachella) [reply deepinv/tests/test_models.py]: done

File: deepinv/tests/test_optim.py (hunk 6 above)
Comment (mh-nguyen712) [reply deepinv/tests/test_optim.py:1025]: This test takes quite long time to finish so we risk to increase the test time...

File: deepinv/models/ram.py (hunk 8 above)
Comment (tachella) [reply deepinv/models/ram.py:189]: I've added `test_ram_scale` to check for this, then I found many bugs with RAM which I fixed

File: deepinv/tests/test_optim.py (hunk 5 above)
Comment (tachella) [reply deepinv/tests/test_optim.py]: I'm only testing batch size 2 for speed, but I left the option with a `parametrize`

File: deepinv/models/ram.py [hunk 14]
```diff
@@ -246,33 +245,47 @@ def forward(self, y, physics=None, sigma=None, gain=None):
...
                 "Either physics, sigma or gain must be provided to the RAM model."
             )
 
-        if physics is None:
-            gain = self.gain_threshold if gain is None else gain
-            sigma = self.sigma_threshold if sigma is None else sigma
+        if isinstance(y, TensorList):
+            max_val = y[0].abs().amax(dim=(1, 2, 3), keepdim=False)
+        else:
+            max_val = y.abs().amax(dim=(1, 2, 3), keepdim=False)
```
Comment (Andrewwango) [inline deepinv/models/ram.py]: why `dim=(1,2,3)`? do you mean reducing over all dims that aren't the batch dim? Then you should do between 1 and ndim. `(1,2,3)` will fail for e.g. multicoil MRI data of shape `BCNHW`

Comment (Andrewwango) [reply deepinv/models/ram.py]: secondly, what is the reasoning behind switching from max_val calculated on ATy to calculated on y?

File: deepinv/models/ram.py [hunk 15]
```diff
@@ -246,33 +245,47 @@ def forward(self, y, physics=None, sigma=None, gain=None):
...
 
-            physics = dinv.physics.Denoising(
-                noise_model=dinv.physics.PoissonGaussianNoise(sigma=sigma, gain=gain),
-            )
+        rescale_val = torch.where(
```
Comment (Andrewwango) [inline deepinv/models/ram.py:271]: I think you will need to add an inline comment here, as it is not immediately obvious what is going on

File: deepinv/models/ram.py [hunk 16]
```diff
@@ -246,33 +245,47 @@ def forward(self, y, physics=None, sigma=None, gain=None):
...
+        else:
+            y = y / rescale_val.view([y.shape[0]] + [1] * (y.ndim - 1))
 
-        max_val = x_temp.abs().max()
-        rescale_val = 1.0 if max_val > 5 * self.sigma_threshold else max_val
-        y = y / rescale_val
+        if physics is None:
+            physics = dinv.physics.Denoising(noise_model=dinv.physics.ZeroNoise())
+
+        x_temp = physics.A_adjoint(y)
```
Comment (Andrewwango) [inline deepinv/models/ram.py:286]: What is the purpose of this x_temp? Can we remove this and use x_in below?

File: deepinv/models/ram.py [hunk 17]
```diff
@@ -246,33 +245,47 @@ def forward(self, y, physics=None, sigma=None, gain=None):
...
 
         sigma, gain = self.obtain_sigma_gain(
-            physics=physics, y=y, sigma=sigma, gain=gain, rescale_val=rescale_val
+            physics, sigma, gain, rescale_val, device=y.device
```
Comment (Andrewwango) [inline deepinv/models/ram.py]: Why removing the kwargs? it's safer

File: deepinv/models/ram.py [hunk 18]
```diff
@@ -292,54 +305,53 @@ def forward(self, y, physics=None, sigma=None, gain=None):
...
         :param float, torch.Tensor sigma: Gaussian noise level. If None, will be set to the threshold.
         :param float, torch.Tensor gain: Poisson noise level. If None, will be set to the threshold.
-        :param float rescale_val: Rescale value to apply to the sigma and gain values.
+        :param torch.Tensor rescale_val: Rescale value to apply to the sigma and gain values.
+        :param str, torch.device device: Device to which the sigma and gain values should be moved.
         """
-        if hasattr(physics, "noise_model"):
-            if sigma is not None or gain is not None:
-                warn(
-                    "noise_model specified in physics. Parameters passed to sigma or gain will be ignored."
```
Comment (Andrewwango) [inline deepinv/models/ram.py:303]: Surely we will need to keep this warning somewhere, when a user does `model(y, Denoising(GaussianNoise(0.1)), sigma=0.2)`?

File: deepinv/tests/test_models.py [hunk 19]
```diff
@@ -856,6 +856,55 @@ def __len__(self):
     assert psnr(x_init, x) < psnr(x_hat, x)
 
 
+@pytest.mark.parametrize("use_physics", [True, False])
+@pytest.mark.parametrize("scale", [-5, 5])
```
Comment (Andrewwango) [inline deepinv/tests/test_models.py]: minor point, I would use 1e-5, 1e5 here to avoid any confusion

File: deepinv/models/ram.py (hunk 14 above)
Comment (tachella) [reply deepinv/models/ram.py]: the idea is that noise levels are defined on the measurement space and not on the signal space, so scaling should happen on the measurement space

File: deepinv/models/ram.py (hunk 18 above)
Comment (tachella) [reply deepinv/models/ram.py:303]: If the user specifies a sigma, I believe it is because they want to use that sigma (and not the sigma hidden in the physics). I think that the previous behavior, i.e. ignoring the user-specified sigma, was more confusing.

File: deepinv/models/ram.py (hunk 16 above)
Comment (tachella) [reply deepinv/models/ram.py:286]: unfortunately we cannot get rid of it, since we need it to guess the size of the image to then use `PhysicsCropper`

File: deepinv/tests/test_models.py (hunk 19 above)
Comment (tachella) [reply deepinv/tests/test_models.py]: done

File: deepinv/models/ram.py (hunk 17 above)
Comment (tachella) [reply deepinv/models/ram.py]: I added the kwargs back

File: deepinv/models/ram.py (hunk 15 above)
Comment (tachella) [reply deepinv/models/ram.py:271]: I've added a comment - the goal is to rescale only the batch elements that have very large/small values

File: deepinv/models/ram.py (hunk 14 above)
Comment (tachella) [reply deepinv/models/ram.py]: fixed the dim issue

File: deepinv/models/ram.py (hunk 18 above)
Comment (Andrewwango) [reply deepinv/models/ram.py:303]: Fair enough but I would add a warning in case

Comment (tachella) [reply deepinv/models/ram.py:303]: Ok, I've added a warning
